dvc/remote/base.py

```diff
@@ -381,18 +382,105 @@ def _save(self, path_info, checksum):
             return
         self._save_file(path_info, checksum)
 
+    @contextmanager
+    def transfer_context(self):
+        yield None
+
+    def upload(self, from_infos, to_infos, names=None, no_progress_bar=False):
+        if not hasattr(self, "_upload"):
+            raise RemoteActionNotImplemented("upload", self.scheme)
+        names = self._verify_path_args(to_infos, from_infos, names)
+
+        with self.transfer_context() as ctx:
+            for from_info, to_info, name in zip(from_infos, to_infos, names):
+                if to_info.scheme != self.scheme:
+                    raise NotImplementedError
+
+                if from_info.scheme != "local":
+                    raise NotImplementedError
+
+                msg = "Uploading '{}' to '{}'"
+                logger.debug(msg.format(from_info, to_info))
+
+                if not name:
+                    name = from_info.name
+
+                if not no_progress_bar:
+                    progress.update_target(name, 0, None)
+
+                try:
+                    self._upload(
+                        from_info.fspath,
+                        to_info,
+                        name=name,
+                        ctx=ctx,
+                        no_progress_bar=no_progress_bar,
+                    )
+                except Exception:
+                    msg = "failed to upload '{}' to '{}'"
+                    logger.exception(msg.format(from_info, to_info))
+                    continue
+
+                if not no_progress_bar:
+                    progress.finish_target(name)
+
     def download(
         self,
         from_infos,
         to_infos,
         no_progress_bar=False,
```

Comment:
can we make it "progress_bar" - so that we don't have code like `not no_progress_bar` with a double negation - hard to read

Reply:
I agree. This will, however, touch some more places. I plan to fix this separately late.

Reply:
👍